Svc/FpySequencer/FpySequencerValidationState.cpp

```diff
@@ -0,0 +1,181 @@
+        if (deserStatus != Fw::FW_SERIALIZE_OK) {
+            this->log_WARNING_HI_FileReadDeserializeError(m_sequenceFilePath, static_cast<I32>(deserStatus),
+                                                          m_sequenceBuffer.getBuffLeft(),
+                                                          m_sequenceBuffer.getBuffLength());
+            return false;
+        }
+    }
+
+    // deser statements
+    for (U16 statementIdx = 0; statementIdx < m_sequenceObj.getheader().getstatementCount(); statementIdx++) {
+        // deser statement
+        deserStatus = m_sequenceBuffer.deserialize(m_sequenceObj.getstatements()[statementIdx]);
+        if (deserStatus != Fw::FW_SERIALIZE_OK) {
+            this->log_WARNING_HI_FileReadDeserializeError(m_sequenceFilePath, static_cast<I32>(deserStatus),
+                                                          m_sequenceBuffer.getBuffLeft(),
+                                                          m_sequenceBuffer.getBuffLength());
+            return false;
+        }
+    }
+
+    // read footer bytes (but don't include in CRC)
+    // TODO i think if we read the crc into the crc, it has the mathematical property
+    // that the new crc will be 0. why not check this instead of comparing crcs?
+    readStatus = readBytes(sequenceFile, Fpy::Footer::SERIALIZED_SIZE, false);
+    if (!readStatus) {
+        return false;
+    }
+
+    deserStatus = m_sequenceBuffer.deserialize(m_sequenceObj.getfooter());
+    if (deserStatus != Fw::FW_SERIALIZE_OK) {
+        this->log_WARNING_HI_FileReadDeserializeError(m_sequenceFilePath, static_cast<I32>(deserStatus),
+                                                      m_sequenceBuffer.getBuffLeft(), m_sequenceBuffer.getBuffLength());
+        return false;
+    }
+
+    // need this for some reason to "finalize" the crc TODO get an explanation on this
+    this->m_computedCRC = ~this->m_computedCRC;
+
+    if (this->m_computedCRC != m_sequenceObj.getfooter().getcrc()) {
+        this->log_WARNING_LO_WrongCRC(m_sequenceObj.getfooter().getcrc(), m_computedCRC);
+        return false;
+    }
+
+    return true;
+}
+
+// reads some bytes from the open file into the m_sequenceBuffer.
+// return true if successful
+bool FpySequencer::readBytes(Os::File& file, FwSizeType readLen, bool updateCRC) {
+    FW_ASSERT(file.isOpen());
+    // this has to be fwsignedsizetype cuz that's what file.read takes
+    // it also has to be declared a var because file.read must take a ref
+    FwSignedSizeType expectedReadLen = readLen;
+
+    const FwSizeType capacity = m_sequenceBuffer.getBuffCapacity();
+
+    // if this asserts, then you need to give the sequencer more buffer memory. pass in a bigger number
+    // to fpySeq.allocateBuffer(). This is usually done in topology setup CPP
+    FW_ASSERT(static_cast<FwSignedSizeType>(capacity) >= expectedReadLen, static_cast<FwAssertArgType>(capacity),
```

Comment:
Always upcast to the larger ranged type.
```
FW_ASSERT(capacity >= static_cast<FwSizeType>(expectedReadLen), 
```
While probably ok, your cast could overflow.  This is easier to contend with.